04-onion-routing.md

```diff
@@ -755,59 +763,143 @@ Note that these error messages are not reliable, as they are not placed on-chain
 due to the possibility of hop failure.
 
 Intermediate hops store the shared secret from the forward path and reuse it to
-obfuscate any corresponding return packet during each hop.
+authenticate and obfuscate any corresponding return packet during each hop.
 In addition, each node locally stores data regarding its own sending peer in the
 route, so it knows where to return-forward any eventual return packets.
-The node generating the error message (_erring node_) builds a return packet
+
+The origin node signals to each node on the path that it supports the error
+format that is detailed below by setting the `resolution_format` tlv record in
+the forward packet to `1`. It must make sure that each node supports this format
+by observing the node feature `option_fat_error`.
+
+## Erring node
+
+The node generating the error message builds a return packet
 consisting of the following fields:
 
 1. data:
-   * [`32*byte`:`hmac`]
    * [`u16`:`failure_len`]
    * [`failure_len*byte`:`failuremsg`]
    * [`u16`:`pad_len`]
    * [`pad_len*byte`:`pad`]
+   * [`27*9*byte`:`payloads`]
+   * [`378*32*byte`:`hmacs`]
+
+Where `failuremsg` is defined below, and `pad` are the extra bytes used to
+conceal length.
+
+The field `payloads` contains a per-hop payload. The erring node puts its
+payload at the start of this array and zeroes out the rest. The size of the
+field is based on the maximum number of hops in a route (27) and the payload
+size (9 bytes).
+
+The per-hop payload consists of the following fields:
+1. data:
+   * [`byte`:`payload_type`]
+   * [`u64`:`hold_time_ms`]
+
+Where `payload_type` indicates whether this hop is an intermediate hop (value 0)
+or the source of the error (value 1). The erring node sets this to 1. Via
+`hold_time_ms` each hops reports the time that it held on to the htlc. The
+sender can use this information to identify slow nodes and avoid them for future
+payment attempts. 
 
-Where `hmac` is an HMAC authenticating the remainder of the packet, with a key
-generated using the above process, with key type `um`, `failuremsg` as defined
-below, and `pad` as the extra bytes used to conceal length.
+The field `hmacs` contains authentication codes for each hop, with a `um` type
+key generated using the above process. The size of the field is based on the
+maximum number of hops in a route (27). Each hop adds 27 hmacs, one for each
+possible position that the hop could be at in the path. This is necessary
+because only the sender knows the position of each hop in the path. At each step
+back, some of the hmacs can be pruned. Pruning leads to a space requirement
+of (27 + 28) / 2 = 378 hmacs.
 
-The erring node then generates a new key, using the key type `ammag`.
-This key is then used to generate a pseudo-random stream, which is in turn
-applied to the packet using `XOR`.
+The layout of the `hmacs` field shown below. The actual format is much longer,
```

Comment:
Yes maybe. But this PR has to be rewritten at some point anyway, because it's currently replacing the legacy failure message format. Will do it once we've decided which documentation style is preferable for this change.